todo/src/util/dnd.js

```diff
@@ -0,0 +1,210 @@
+
+  CONSTANTS = {
+    FOCUS_COLOR: '1px solid #000a',
+    FOCUS_ZINDEX: '1000'
+  }
+
+  constructor(element, dropzoneList) {
+    this.element = element;
+    this.dropzoneList = dropzoneList;
+    this.addDraggableStyle();
+    this.addDraggableEvent();
+  }
+
+  addDraggableStyle() {
+    this.element.style.cursor = 'pointer';
+  }
+
+  addDraggableEvent() {
+    this.downEvent = this.createDownEvent();
+    addEvent(this.element, 'mousedown', this.downEvent);
+  }
+
+  createDownEvent() {
+    return downEvent => {
+      downEvent.preventDefault();
+
+      const elementRect = this.element.getBoundingClientRect();
+      const basePos = {
+        x: downEvent.clientX - elementRect.x,
+        y: downEvent.clientY - elementRect.y
+      };
+
+      this.createCloneNode();
+      this.focus();
+
+      const mouseMoveEvent = this.createMoveEvent(downEvent, basePos);
+      addEvent(document, 'mousemove', mouseMoveEvent);
+      addEvent(this.element, 'mouseup', this.createMouseUpEvent(mouseMoveEvent, basePos));
+    };
+  }
+
+  createMoveEvent(downEvent, basePos) {
+    return moveEvent => {
+      moveEvent.preventDefault();
+      const pos = {
+        xPos: moveEvent.clientX - basePos.x,
+        yPos: moveEvent.clientY - basePos.y
+      };
+      this.moveToAt(pos);
+
+      this.dropzoneList.forEach(dropzone => {
+        if (dropzone.contains(downEvent)) {
+          dropzone.fire('drag', {
+            target: this.element,
+            event: moveEvent
+          });
+        }
+      });
+    };
+  }
+
+  createMouseUpEvent(moveEvent) {
+    return upEvent => {
+      upEvent.preventDefault();
+      const dropzone = this.getDropzone(upEvent);
+      if (dropzone instanceof Dropzone) {
+        dropzone.fire('drop', {
+          target: this.element,
+          isContain: dropzone.contains(this.element.getBoundingClientRect())
+        });
+      }
+
+      this.element.removeEventListener('mouseup', upEvent);
+      document.removeEventListener('mousemove', moveEvent);
+      this.rollbackNode();
+    };
+  }
+
+  moveToAt({xPos, yPos}) {
+    this.element.style.left = `${xPos}px`;
+    this.element.style.top = `${yPos}px`;
+  }
+
+  createCloneNode() {
+    this.cloneNode = this.element.cloneNode(true);
+    this.cloneNode.style.opacity = 0.3;
+    this.element.parentNode.insertBefore(this.cloneNode, this.element);
+    this.element.style.position = 'absolute';
+  }
+
+  rollbackNode() {
+    this.element.parentNode.removeChild(this.element);
+    this.element = this.cloneNode;
+    this.addDraggableEvent();
+    this.element.style.opacity = '';
+  }
+
+  focus() {
+    this.element.style.border = this.CONSTANTS.FOCUS_COLOR;
+    this.element.style.zIndex = this.CONSTANTS.FOCUS_ZINDEX;
+  }
+
+  getDropzone(event) {
+    if (this.dropzoneList.length === 0) {
+      return null;
+    }
+
+    return this.dropzoneList.find(dropzone => dropzone.contains({
+      x: event.clientX,
+      y: event.clientY
+    })) ?? null;
+  }
+}
```

Comment:
왼쪽 피연산자의 값이 `undefined` 일 때 `null`을 반환하기 위해 널 병합 연산자를 사용하신것 같습니다.
함수는 값을 반환하지 않으면 기본적으론 `undefined`를 반환하기 때문에 아래처럼 작성하면 Dropzone을 찾지 못한 모든 경우에 일률적으로 `undefined`를 반환할 수 있을 것 같습니다.

```suggestion
  getDropzone(event) {
    if (this.dropzoneList.length === 0) {
      return;
    }

    return this.dropzoneList.find(dropzone => dropzone.contains({
      x: event.clientX,
      y: event.clientY
    }));
  }
}
```

Reply:
`undefined` or `null`을 검사하기 위해 널 병합 연산자를 사용하는데, 널을 반환하는건 지금 보니 이상하네요.
- 추가적으로 `length === 0` 검사도 안해도 괜찮겠네요!